src/resolve/resolved-mdns-browse-services.c

```diff
@@ -0,0 +1,697 @@
+        int r;
+
+        assert(userdata);
+
+        ss = mdns_service_subscriber_ref(userdata);
+
+        r = dns_query_new(ss->m, &q, ss->question_utf8, ss->question_idna, NULL, ss->ifindex, ss->flags);
+        if (r < 0)
+                goto finish;
+
+        q->complete = mdns_browse_service_query_complete;
+        q->varlink_request = varlink_ref(ss->link);
+        varlink_set_userdata(ss->link, q);
+
+        r = dns_query_go(q);
+        if (r < 0)
+                goto finish;
+
+        /* RFC6762 5.2
+         * The intervals between successive queries MUST increase by at least a factor of two.
+         * When the interval between queries reaches or exceeds 60 minutes,perform
+         * subsequent queries at a steady-state rate of one query per hour */
+        if (ss->delay == 0) {
+                ss->delay++;
+                /* First query is sent wihtout SD_RESOLVED_NO_CACHE to fetch answers already in cache.
+                 * Set SD_RESOLVED_NO_CACHE to make all subsequent queries go to the network. */
+                ss->flags |= SD_RESOLVED_NO_CACHE;
+        }
+        else
+                ss->delay = ss->delay < 2048 ? ss->delay * 2 : 3600;
+
+        r = event_reset_time_relative(
+                ss->m->event, &ss->schedule_event,
+                CLOCK_BOOTTIME, (ss->delay * USEC_PER_SEC),
+                0, mdns_next_query_schedule,
+                ss, 0, "mdns-next-query-schedule", true);
+        if (r < 0)
+                goto finish;
+
+        TAKE_PTR(q);
+
+        return 0;
+
+finish:
+        return log_error_errno(r, "Failed to schedule mDNS query, %m");
+}
+
+void mdns_ss_reset(Manager *m) {
+        int r;
+
+        if (!(m && m->dns_service_subscriber))
+                return;
+
+        mDnsServiceSubscriber *ss = m->dns_service_subscriber;
+
+        ss->delay = 0;
+
+        r = event_reset_time_relative(
+                ss->m->event, &ss->schedule_event,
+                CLOCK_BOOTTIME, (ss->delay * USEC_PER_SEC),
+                0, mdns_next_query_schedule,
+                ss, 0, "mdns-next-query-schedule", true);
+        if (r < 0)
+                log_error_errno(r, "Failed to reset mdns service subscriber, %m");
+
+        return;
+}
+
+int mdns_subscribe_browse_service(
+                Manager *m,
+                Varlink *link,
+                const char *domain,
+                const char *name,
+                const char *type,
+                const char *ifname,
+                const uint64_t token) {
+
+        _cleanup_(mdns_service_subscriber_unrefp) mDnsServiceSubscriber *ss = NULL;
+        _cleanup_(dns_question_unrefp) DnsQuestion *question_idna = NULL, *question_utf8 = NULL;
+        int ifindex;
+        int r;
+
+        assert(m);
+        assert(link);
+
+        ifindex = rtnl_resolve_ifname(NULL, ifname);
+        if (ifindex <= 0)
+                return varlink_error_invalid_parameter(link, JSON_VARIANT_STRING_CONST("ifname"));
```

Comment:
this should not take an ifname, but an ifindex.

Reply:
Changed to take ifindex